File: packages/claude-plugins/worktree/scripts/worktree-create.ts
#!/usr/bin/env tsx
import { readFileSync, writeFileSync } from 'node:fs'
import { join } from 'node:path'
import {
    readRegistry,
    writeRegistry,
    findFreeSlot,
    findSlotByIssue,
    findSlotByBranch,
    allocateSlot
} from './lib/registry.js'
import { readSlots, getSlotConfig } from './lib/slots.js'
import {
    getTemplateFiles,
    getOutputFileName,
    processTemplate,
    readEnvFile,
    findMainEnvFiles
} from './lib/env.js'
import {
    getRepoName,
    getMainBranch,
    branchExists,
    createBranch,
    createWorktree
} from './lib/git.js'
import { fetchIssue, createBranchName, isGithubRepo } from './lib/github.js'
import { checkPorts, extractPortsFromSlot } from './lib/ports.js'
import {
    getWorktreesDir,
    getConfigDir,
    getWorktreePath,
    getRelativePath,
    worktreeNameFromBranch,
    configExists
} from './lib/paths.js'

interface CreateOptions {
    dryRun?: boolean
}

interface CreateResult {
    success: boolean
    slot: number
    branch: string
    worktreePath: string
    relativePath: string
    portWarnings: string[]
    envWarnings: string[]
}

export async function create(
    mainRepoDir: string,
    input: string | number,
    options: CreateOptions = {}
): Promise<CreateResult> {
    const worktreesDir = getWorktreesDir(mainRepoDir)
    const configDir = getConfigDir(worktreesDir)

    // Check config exists
    if (!configExists(worktreesDir)) {
        throw new Error('No worktree config found. Run /worktree:init first.')
    }

    // Read registry and slots
    const registry = readRegistry(worktreesDir)
    const slotsConfig = readSlots(worktreesDir)

    // Determine branch name
    let branch: string
    let issueNumber: number | null = null

    if (typeof input === 'number' || /^\d+$/.test(String(input))) {
        issueNumber = typeof input === 'number' ? input : parseInt(input, 10)

        // Check if already exists
        const existing = findSlotByIssue(registry, issueNumber)
        if (existing) {
            throw new Error(`Worktree for issue #${issueNumber} already exists (slot ${existing.slot})`)
        }

        // Fetch issue from GitHub
        const issue = fetchIssue(issueNumber)
        if (!issue) {
            throw new Error(`GitHub issue #${issueNumber} not found`)
        }
        branch = createBranchName(issueNumber, issue.title)
        console.log(`Issue #${issueNumber}: ${issue.title}`)
    } else {
        branch = String(input)
        // Check if branch already has worktree
        const existing = findSlotByBranch(registry, branch)
        if (existing) {
            throw new Error(`Worktree for branch "${branch}" already exists (slot ${existing.slot})`)
        }
    }

    // Find free slot
    const freeSlot = findFreeSlot(registry)
    if (!freeSlot) {
        throw new Error(`All ${registry.slotCount} slots in use. Remove a worktree first.`)
    }

    const slotConfig = getSlotConfig(slotsConfig, freeSlot.slot)
    if (!slotConfig) {
        throw new Error(`Slot ${freeSlot.slot} not defined in slots.yaml`)
    }

    // Check ports
    const ports = extractPortsFromSlot(slotConfig)
    const portStatus = await checkPorts(ports)
    const portWarnings: string[] = []
    for (const [port, inUse] of portStatus) {
        if (inUse) {
            portWarnings.push(`Port ${port} appears to be in use`)
        }
    }

    const worktreeName = worktreeNameFromBranch(branch)
    const worktreePath = getWorktreePath(worktreesDir, worktreeName)
    const relativePath = getRelativePath(mainRepoDir, worktreePath)

    console.log(`\nCreating worktree:`)
    console.log(`  Branch: ${branch}`)
    console.log(`  Slot: ${freeSlot.slot}`)
    console.log(`  Path: ${relativePath}`)

    if (options.dryRun) {
        console.log('\n[DRY RUN] Would create worktree and process templates')
        return {
            success: true,
            slot: freeSlot.slot,
            branch,
            worktreePath,
            relativePath,
            portWarnings,
            envWarnings: []
        }
    }

    // Create branch if needed
    const mainBranch = getMainBranch(mainRepoDir)
    if (!branchExists(branch, mainRepoDir)) {
        console.log(`Creating branch from ${mainBranch}...`)
        createBranch(branch, mainBranch, mainRepoDir)
    }

    // Create worktree
    console.log('Creating worktree...')
    createWorktree(worktreePath, branch, mainRepoDir)

    // Process templates
    console.log('Processing .env templates...')
    const templateFiles = getTemplateFiles(configDir)
    const mainEnvFiles = findMainEnvFiles(mainRepoDir)
    const mainEnvVars = new Map<string, string>()
    for (const envFile of mainEnvFiles) {
        const vars = readEnvFile(envFile)
        for (const [k, v] of vars) {
            mainEnvVars.set(k, v)
        }
    }

    const envWarnings: string[] = []
    for (const templatePath of templateFiles) {
        const templateContent = readFileSync(templatePath, 'utf-8')
        const { content, warnings } = processTemplate(templateContent, slotConfig, mainEnvVars)
        envWarnings.push(...warnings)

        const outputName = getOutputFileName(templatePath)
        const outputPath = join(worktreePath, outputName)
        writeFileSync(outputPath, content)
        console.log(`  ✓ Created ${outputName}`)
    }

    // Update registry
    allocateSlot(registry, freeSlot.slot, issueNumber, branch, worktreeName)
    writeRegistry(worktreesDir, registry)

    return {
        success: true,
        slot: freeSlot.slot,
        branch,
        worktreePath,
        relativePath,
        portWarnings,
        envWarnings
    }
}

// CLI interface
if (import.meta.url === `file://${process.argv[1]}`) {
    const args = process.argv.slice(2)
    const dryRun = args.includes('--dry-run')
    const input = args.find(a => !a.startsWith('--'))

    if (!input) {
        console.error('Usage: worktree-create <issue-number|branch-name> [--dry-run]')
        process.exit(1)
    }

    create(process.cwd(), input, { dryRun })
        .then(result => {
            console.log('\n✅ Worktree created!')
            console.log(`\nSlot ${result.slot} assigned`)
            console.log(`Path: ${result.relativePath}`)

            if (result.portWarnings.length > 0) {
                console.log('\n⚠️  Port warnings:')
                result.portWarnings.forEach(w => console.log(`  - ${w}`))
            }

            if (result.envWarnings.length > 0) {
                console.log('\n⚠️  Template warnings:')
                result.envWarnings.forEach(w => console.log(`  - ${w}`))
            }

            console.log('\nNext steps:')
            console.log(`  cd ${result.relativePath}`)
            console.log('  pnpm install  # or npm install')
            console.log('  pnpm dev      # start development')
        })
        .catch(err => {
            console.error(`\n❌ Error: ${err.message}`)
            process.exit(1)
        })
}

File: packages/towles-tool/worktree/skills/worktree_create.py
#!/usr/bin/env python3
"""Create a new worktree with slot allocation."""
from __future__ import annotations

import sys
from dataclasses import dataclass
from pathlib import Path
from typing import Optional

from lib.registry import (
    read_registry,
    write_registry,
    find_free_slot,
    find_slot_by_issue,
    find_slot_by_branch,
    allocate_slot,
)
from lib.slots import read_slots, get_slot_config
from lib.env import (
    get_template_files,
    get_output_filename,
    process_template,
    read_env_file,
    find_main_env_files,
)
from lib.git import (
    get_main_branch,
    branch_exists,
    create_branch,
    create_worktree,
)
from lib.github import fetch_issue, create_branch_name
from lib.ports import check_ports, extract_ports_from_slot
from lib.paths import (
    get_worktrees_dir,
    get_config_dir,
    get_worktree_path,
    get_relative_path,
    worktree_name_from_branch,
    config_exists,
)


@dataclass
class CreateResult:
    success: bool
    slot: int
    branch: str
    worktree_path: str
    relative_path: str
    port_warnings: list[str]
    env_warnings: list[str]


def create(
    main_repo_dir: Path,
    input_arg: str | int,
    dry_run: bool = False,
) -> CreateResult:
    """Create a new worktree with slot allocation."""
    worktrees_dir = get_worktrees_dir(main_repo_dir)
    config_dir = get_config_dir(worktrees_dir)

    # Check config exists
    if not config_exists(worktrees_dir):
        raise RuntimeError("No worktree config found. Run /worktree:init first.")

    # Read registry and slots
    registry = read_registry(worktrees_dir)
    slots_config = read_slots(worktrees_dir)

    # Determine branch name
    branch: str
    issue_number: Optional[int] = None

    input_str = str(input_arg)
    if input_str.isdigit():
        issue_number = int(input_str)

        # Check if already exists
        existing = find_slot_by_issue(registry, issue_number)
        if existing:
            raise RuntimeError(f"Worktree for issue #{issue_number} already exists (slot {existing.slot})")

        # Fetch issue from GitHub
        issue = fetch_issue(issue_number)
        if not issue:
            raise RuntimeError(f"GitHub issue #{issue_number} not found")
        branch = create_branch_name(issue_number, issue.title)
        print(f"Issue #{issue_number}: {issue.title}")
    else:
        branch = input_str
        # Check if branch already has worktree
        existing = find_slot_by_branch(registry, branch)
        if existing:
            raise RuntimeError(f'Worktree for branch "{branch}" already exists (slot {existing.slot})')

    # Find free slot
    free_slot = find_free_slot(registry)
    if not free_slot:
        raise RuntimeError(f"All {registry.slot_count} slots in use. Remove a worktree first.")

    slot_config = get_slot_config(slots_config, free_slot.slot)
    if not slot_config:
        raise RuntimeError(f"Slot {free_slot.slot} not defined in slots.toml")

    # Check ports
    ports = extract_ports_from_slot(slot_config)
    port_status = check_ports(ports)
    port_warnings = [f"Port {port} appears to be in use" for port, in_use in port_status.items() if in_use]

    worktree_name = worktree_name_from_branch(branch)
    worktree_path = get_worktree_path(worktrees_dir, worktree_name)
    relative_path = get_relative_path(main_repo_dir, worktree_path)

    print("\nCreating worktree:")
    print(f"  Branch: {branch}")
    print(f"  Slot: {free_slot.slot}")
    print(f"  Path: {relative_path}")

    if dry_run:
        print("\n[DRY RUN] Would create worktree and process templates")
        return CreateResult(
            success=True,
            slot=free_slot.slot,
            branch=branch,
            worktree_path=str(worktree_path),
            relative_path=relative_path,
            port_warnings=port_warnings,
            env_warnings=[],
        )

    # Create branch if needed
    main_branch = get_main_branch(main_repo_dir)
    if not branch_exists(branch, main_repo_dir):
        print(f"Creating branch from {main_branch}...")
        create_branch(branch, main_branch, main_repo_dir)

    # Create worktree
    print("Creating worktree...")
    create_worktree(worktree_path, branch, main_repo_dir)

    # Process templates
    print("Processing .env templates...")
    template_files = get_template_files(config_dir)
    main_env_files = find_main_env_files(main_repo_dir)
    main_env_vars: dict[str, str] = {}
    for env_file in main_env_files:
        main_env_vars.update(read_env_file(env_file))

    env_warnings: list[str] = []
    for template_path in template_files:
        template_content = template_path.read_text()
        content, warnings = process_template(template_content, slot_config, main_env_vars)
        env_warnings.extend(warnings)

        output_name = get_output_filename(template_path)
        output_path = worktree_path / output_name
        output_path.write_text(content)
        print(f"  ✓ Created {output_name}")

    # Update registry
    allocate_slot(registry, free_slot.slot, issue_number, branch, worktree_name)
    write_registry(worktrees_dir, registry)

    return CreateResult(
        success=True,
        slot=free_slot.slot,
        branch=branch,
        worktree_path=str(worktree_path),
        relative_path=relative_path,
        port_warnings=port_warnings,
        env_warnings=env_warnings,
    )


def main() -> None:
    args = sys.argv[1:]
    dry_run = "--dry-run" in args
    args = [a for a in args if not a.startswith("--")]

    if not args:
        print("Usage: worktree_create.py <issue-number|branch-name> [--dry-run]", file=sys.stderr)
        sys.exit(1)

    try:
        result = create(Path.cwd(), args[0], dry_run=dry_run)
        print("\n✅ Worktree created!")
        print(f"\nSlot {result.slot} assigned")
        print(f"Path: {result.relative_path}")

        if result.port_warnings:
            print("\n⚠️  Port warnings:")
            for w in result.port_warnings:
                print(f"  - {w}")

        if result.env_warnings:
            print("\n⚠️  Template warnings:")
            for w in result.env_warnings:
                print(f"  - {w}")

        print("\nNext steps:")
        print(f"  cd {result.relative_path}")
        print("  pnpm install  # or npm install")
        print("  pnpm dev      # start development")

    except Exception as e:
        print(f"\n❌ Error: {e}", file=sys.stderr)
        sys.exit(1)


if __name__ == "__main__":
    main()
